Rewriting RzNum parser and calculator (#4326)
rizinorg/rizin: doc/math.md

    $f    fail-jump address             $j   true-jump address
    $Ja   analysis jump address
    $l    last instruction length       $M   map count
    $MM   address space size            $m   memory size
    $O    current opcode address        $o   opcode size
    $p    PID                           $P   parent PID
    $r    return value register         $s   stack pointer
    $v    value                         $w   word size

An unknown `$`-name is a parse error (since the rule is a closed
list), not an evaluator error.

Variable bindings
-----------------

`x = y` and `let x = y` bind the identifier `x` to the value of `y`
and evaluate to that value. The two forms are identical; `let` is an
optional keyword with no separate scoping. The binding is visible to
the rest of the current expression, so a later part can read it back:

    (x = 7) + x              -> 14
    (a = 2) * (a + 1)        -> 6
    (c = 1) + (c = 9) + c    -> 19    (reassignment updates the binding)

A local binding shadows a host variable (flag / special variable)
of the same name. Bound values keep their kind, so a bound big
number stays exact. Reserved words (`mod`, `log`, `le`, `be`,

> `sdiv`, `smod`, `sar`) cannot be assigned.

Reviewer: That list differs with the reserved words above.
Better make one table with reserved words and reference it.

With the statement separator `;` (see below) bindings can be
introduced and reused without the parenthesis trick:

    let radius = 5; let area = radius * radius; area    -> 25
    bla = crc(0, 4); 0:be32 + bla

Persistence across evaluations is opt-in. By default a binding lives
for a single evaluation. If the caller supplies a persistent variable
store (`RzNumMathOptions.vars`, created with `rz_num_value_store_new()`),
bindings written by one evaluation remain visible to later ones that
share the store. The `%` shell command uses such a store hung off the
core, so a variable bound in one `%` invocation is visible to the
next:

> 

Reviewer: ```suggestion
share the store. The `%` shell command uses such a store, so a variable bound in one `%` invocation is visible to the
next:

```

    [0x0]> % bla = crc(0, 4)
    [0x0]> % 0:be32 + bla     <- 'bla' still bound here

Statement sequencing
--------------------

Several expressions can be chained with `;`. Each is evaluated in
order (so bindings made by earlier statements are visible to later
ones) and the value of the whole is the value of the last statement.
A trailing `;` is allowed.

    1 + 2; 3 + 4             -> 7
    x = 10; y = 20; x + y    -> 30

>     a = 3; b = a * a; a + b  -> 12

Reviewer: ```suggestion
    a = 3; b = a * a; a + b;  -> 12
```

So there is an example for that case.

Operators
---------

In order from highest precedence to lowest. All operators are
left-associative except exponentiation, which is right-associative.

    Unary prefix:    +x   -x   !x   ~x   ++x   --x

>     Exponent:        x ** y    (right-assoc)

Reviewer: ```suggestion
    Exponent:        x ** y    (right-associative)
```

    Shifts:          x << y   x >> y   x <<< y   x >>> y   x sar y
                     (<<< and >>> are bitwise rotates on the low 64 bits;
                      sar is the sign-propagating arithmetic shift right)
    Bitwise AND:     x & y
    Bitwise XOR:     x ^ y
    Bitwise OR:      x | y
    Comparison:      x < y    x <= y   x > y    x >= y
    Equality:        x == y   x != y
    Conditional:     c ? a : b    (ternary, right-associative)
    Logarithm:       x log y      (computes log_x(y))
    Assignment:      x = y    let x = y    (binds x, evaluates to y)

Notes:

- Division and modulo with a zero right-hand side raise
  `RZ_NUM_ERR_DIV_ZERO` (for both the unsigned `/` `%` and the signed
  `sdiv` `smod`).
- `sdiv`, `smod` and `sar` interpret their operands as two's-complement
  signed 64-bit values. `/`, `%` and `>>` are unsigned. For example
  `-5 sdiv 2` is `-2` while `0xfffffffffffffffb / 2` is a large
  positive number; `0xffffffffffffff00 sar 4` sign-extends to
  `0xfffffffffffffff0` whereas `>> 4` does not.
- Rotations `<<<` and `>>>` always operate on a 64-bit projection of
  their operands; they have no width-independent meaning on an

>   arbitrary-precision big number.

Reviewer: So rotation on an `u32` is like a zero extended `u64`?
And rotation on a `u128` is like a down cast to `u64`?

- `**` and `log` go through `double`. For an exact bignum power use

>   repeated multiplication.

Reviewer: So they also return a double?

    (float); the float decoder handles f16 / f32 / f64 directly;
  * 128-bit reads (`:128` and `:f128`) return a width-128
    bit-vector. Two underlying 64-bit reads are stitched into the
    bit-vector in the requested endianness, so `:128` and `:f128`
    both preserve every bit. RzNum's float kind itself is
    double-precision, so `:f128` does NOT decode to a native
    quad-precision float; the bit-vector carries the raw 128-bit
    IEEE-754 pattern and the host can inspect it, compare two
    quad values bit-for-bit, or feed it to a host function that
    knows how to print quad-precision.

When evaluated through `rz_core_math()` (which backs the `%` shell
command) the suffix dereferences against the active IO layer:

    [0x0]> % 0:le32          (on an ELF)
    hex     0x464c457f       <- little-endian read of 7f 45 4c 46
    [0x0]> % 0:s32           <- same bytes, interpreted signed
    [0x0]> % 0:lef32         <- same bytes, interpreted as a float
    [0x0]> % 0:128           <- 16 bytes as a width-128 bit-vector

When evaluated through the bare `rz_num_math_value()` API with no
IO-read callback wired in, the suffix is parsed but the literal

> address is returned unchanged (parity across every width).

Reviewer: ```suggestion
When evaluated through the bare `rz_num_math_value()` API without
IO layer, the suffix is parsed but the literal
```

Reviewer: Add example.

String-as-bytes literal
-----------------------

A double-quoted string is interpreted as a sequence of bytes packed
little-endian (first byte is the least-significant) into a bit-vector
whose width is `8 * byte-count`. Supported escapes: `\\`, `\"`, `\n`,
`\t`, `\r`, `\0`, `\xHH`.

    "AB"        -> 0x4241        bit-vector, width 16 ('A' in the low byte)
    "ABCD"      -> 0x44434241    bit-vector, width 32
    "\x01\xff"  -> 0xff01        bit-vector, width 16

Because the result is a bit-vector that carries its width, the byte
count is recoverable with `len` (see below) and strings are no longer

> limited to eight bytes:

Reviewer: ```suggestion
count is recoverable with `len`:
```

    len("ABCD")        -> 32     (bits)
    len("ABCD") / 8    -> 4      (bytes)
    len("ABCDEFGHIJ")  -> 80     (a 10-byte string is fine)

Arithmetic on a string value follows bit-vector rules at that width,
so `"AB" + 1` wraps within 16 bits. An empty string `""` is the

> zero-width edge case and evaluates to a `ut64` `0`.

Reviewer: What is with UTF-8 strings?

Also add an example for that string + numerical

Built-in functions
------------------

| Function                       | Arity | Description                            |
|--------------------------------|-------|----------------------------------------|
| `min(a, b)`                    | 2     | smaller operand, kind-preserving       |
| `max(a, b)`                    | 2     | larger operand, kind-preserving        |
| `abs(x)`                       | 1     | absolute value                         |
| `popcount(x)`                  | 1     | number of set bits (works on bignums)  |
| `len(x)`                       | 1     | length in BITS of a bit-vector / big / ut64 (a string literal is a bit-vector, so `len(s)/8` is its byte count) |
| `минимум(a, b)`                | 2     | alias for `min` (Russian)              |
| `максимум(a, b)`               | 2     | alias for `max` (Russian)              |
| `最小(a, b)` / `最大(a, b)`    | 2     | aliases for `min` / `max` (Chinese/Japanese) |
| `دالة(x)`                      | 1     | alias for `abs` (Arabic)               |

> | `Σ(x)`                         | 1     | alias for `popcount` (Greek sigma)     |

Reviewer: Cool :D

pass it through `RzNumMathOptions.funcs`:

    RzNumFuncRegistry *reg = rz_num_func_registry_new();
    rz_num_func_registry_add(reg, "myfn", 1, my_callback, my_userdata);
    RzNumMathOptions opt = { .funcs = reg };
    rz_num_math_value_ex(num, expr, &opt, &out, &err);
    rz_num_func_registry_free(reg);

`rz_num_func_registry_add(reg, name, arity, fn, user)`:

  * `name` is copied as UTF-8, so a function may be registered under
    any identifier the grammar accepts - including a non-ASCII
    (Unicode) name. For example a function registered as `минимум`,
    `合计`, or `δ` is then callable by that exact name:

        rz_num_func_registry_add(reg, "сумма", -1, sum_cb, NULL);
        // expression:  сумма(1, 2, 3)   ->  6

  * `arity` is the exact number of arguments the function requires,
    or `-1` for a variadic function that accepts any number of
    arguments. A call whose argument count does not match a
    non-negative `arity` fails with `RZ_NUM_ERR_NOT_IMPLEMENTED`
    before the callback runs. There is no upper bound on the number
    of arguments a variadic function may receive:

        rz_num_func_registry_add(reg, "сумма", -1, sum_cb, NULL);

>         // сумма(1,2,3,4,5,6,7,8,9,10)   ->  55   (well past 8 args)

Reviewer: ```suggestion
        // сумма(1,2,3,4,5,6,7,8,9,10)   ->  55
```

| `==` `!=` `<` `<=` `>` `>=` | `ut64` 0/1 | unsigned comparison              |

The signed family (`sdiv`/`smod`/`sar`) lifts structurally to RzIL's
signed operators (`/⁺`, `%⁺`, `≫⁺`); rotates, `**`, and `log` have no
direct RzIL pure-op for bit-vectors, so the lift grounds them to a
constant - the evaluator's result is correct, only the symbolic form
is collapsed.

When a bit-vector is combined with a plain `ut64`, the bit-vector's
width wins and the integer is taken modulo that width:

    % 5u8 + 3              -> width 8,  0x08
    % 0xffu8 + 1u8         -> width 8,  0x00   (wraps)
    % 1u8 << 4u8           -> width 8,  0x10
    % 10u8 / 3u8           -> width 8,  0x03

A bit-vector projects to a `ut64` (via `rz_num_value_to_ut64()` or
the legacy `rz_num_math_ut64()`) by taking its low 64 bits.

### Unicode rendering

When the shell is in UTF-8 mode (`scr.utf8=true`), the `%` command
folds the bit-width into the value rows of a bit-vector as a Unicode
subscript - the same notation RzIL uses for bit-vector constants in
its Unicode export - and drops the `0b` prefix from the binary row.
With `scr.utf8=false` the plain ASCII layout is used instead, the

> same way `plf` shows the ASCII layout that `pLf` renders in Unicode.

Reviewer: Too complicated, make sentences shorter.

    [0x0]> e scr.utf8=false
    [0x0]> % 200u8 + 100u8
    hex     0x2c
    binary  0b00101100
    width   8 bits

    [0x0]> e scr.utf8=true
    [0x0]> % 200u8 + 100u8
    hex     0x2c₈
    binary  00101100₈
    width   8 bits

The value rows come first and the width row last in both modes. The
renderer is reachable from C as `rz_num_value_print_ex()` with
`RzNumPrintOptions.utf8` set; `rz_num_value_print()` is the ASCII
default.

Relationship to RzIL
--------------------

RzNum's bit-vector kind and RzIL both build on the same `RzBitVector`
primitive, so the two systems already share a value representation.
The Unicode bit-vector form above is deliberately the same notation
RzIL emits (see `il_export_string_unicode.c`): a hex constant with a

> subscript width. Keeping the two renderers visually consistent means

Reviewer: ```suggestion
RzIL emits (see `il_export_string_unicode.c`). Keeping the two renderers visually consistent means
```

> subscript width. Keeping the two renderers visually consistent means

Reviewer: ```suggestion
RzIL emits (see `il_export_string_unicode.c`): a hex constant with a
subscript width. Keeping the output visually consistent means
```

|----------------|---------------------|----------------|
| `+` `-` `*`    | `add` `sub` `mul`   | `+` `-` `*`    |
| `/` `%`        | `div` `mod`         | `/` `%`        |
| `sdiv` `smod`  | `sdiv` `smod`       | `/⁺` `%⁺` |
| `&` `\|` `^`   | `logand` `logor` `logxor` | `&` `\|` `⊕` |
| `<<` `>>`      | `shiftl` `shiftr`   | `≪` `≫` |
| `sar`          | `shiftr` (sign fill)| `≫⁺` |
| `==`           | `eq`                | `≡`       |
| `<=` (unsigned)| `ule`               | `≦`       |
| `~`            | `lognot`            | `~`            |
| unary `-`      | `neg`               | `−`       |
| `? :` (ternary)| `ite`               | `↠`       |
| `+` `-` `*` `/` (float-pure) | `fadd` `fsub` `fmul` `fdiv` | `(rne x + y)` etc. |
| ternary `? :` (float cond) | `ite` over `is_fzero` | `((f ≡ 0) ↠ alt then)` |

Both of these are now implemented:

  * **A shared bit-vector formatter.** The subscript-width rendering
    used here and the bit-vector case in RzIL's Unicode exporter are
    backed by the same RzUtil helpers, `rz_bv_width_subscript()` and
    `rz_bv_as_unicode_string()`, so the two renderings cannot drift
    apart.

>   * **Lifting an RzNum expression to RzIL.** `rz_num_il_lift()` (and

Reviewer: This can be deleted. Add a reference to the `math-il-lift.md` diretly

condition in `is_fzero` with swapped branches so the lifted form
picks the same branch as the evaluator. A `;`-separated sequence is
still grounded to its final value, since the binding side effects of
earlier statements have no standalone RzIL pure-op form. The
remaining grounded cases are mixed integer-and-float arithmetic
(no implicit cast is invented), function calls, the typed-address
dereference, host variables, and `**` / `log`.

For an exhaustive set of worked examples - simple operators, deeply
nested expressions, the ternary lifting onto RzIL `ite`, host-
registered functions, and every case where the lift has to ground a
sub-expression - see `doc/math-il-lift.md`.

Error categories
----------------

Each evaluation either succeeds (`RZ_NUM_ERR_OK`) or fails with
exactly one category:

| Category                            | When                                       |
|-------------------------------------|--------------------------------------------|
| `RZ_NUM_ERR_PARSE`                  | unbalanced parens, unknown special var, ...|
| `RZ_NUM_ERR_EMPTY`                  | input is empty or whitespace only          |

> | `RZ_NUM_ERR_RESERVED_WORD`          | bare `mod` / `log` / `le` / `be`           |

Reviewer: Refer to "reserved word table"
